### PR TITLE
[#16087]: Backfill shouldn't interfere with scheduled run fix

### LINKED ISSUE #16087: Backfill shouldn't interfere with scheduled run
Example:
DAG structure:
```
default_args = {
    'owner': 'dimon',
    'depends_on_past': False,
    'start_date': datetime(2021, 1, 10)
}

dag = DAG(
    'dummy-dag',
    schedule_interval='21 2 * * *',
    catchup=False,
    default_args=default_args
)

DagContext.push_context_managed_dag(dag)
task1 = BashOperator(task_id='task1', bash_command='echo 1')
task2 = BashOperator(task_id='task2', bash_command='echo 2')
task2 << task1
task3 = BashOperator(task_id='task3', bash_command='echo 3')
```

1. Scheduler creates a DagRun and starts Task Instances processing
2. Scheduler starts “task1” and “task3”
3. At the same time a backfill started for single task - "task3"

Backfill updates scheduler’s DagRun DB table row and scheduler forgets about “task2” and it never gets triggered. Also the DugRun will stuck in the "running" state because backfill will not finish it correctly.
So we have to make a check - if we are starting a backfill with some date, but there is a DagRun in the "running" state for that date already we should not perform a backfill.

### PR DESCRIPTION
**Pull request by dimonchik-suvorov**
closes: #16087

### CODE REVIEW

**Comment by potiuk** [comment]
I think you need to rebase on latest master.

**Comment by houqp** [comment]
@dimonchik-suvorov you should be able to reproduce the CI failures locally using breeze.

File: airflow/jobs/backfill_job.py [hunk 1]
```diff
@@ -689,6 +688,9 @@ def tabulate_tis_set(set_tis: Set[TaskInstance]) -> str:
 
         return err
 
+    def _get_dag_with_subdags(self):
+        return [self.dag] + self.dag.subdags
+
```
Comment (ashb) [inline airflow/jobs/backfill_job.py]:
```suggestion
    def _get_dag_with_subdags(self):
        return [self.dag] + self.dag.subdags
```

Keep it the way it was -- adding a one line function that is only called in one place doesn't make readability any better.

**Review by ashb — CHANGES_REQUESTED** [review]

File: airflow/jobs/backfill_job.py (hunk 1 above)
Comment (dimonchik-suvorov) [reply airflow/jobs/backfill_job.py]: It may help in future refactorings but sure, reverted :)

**Comment by dimonchik-suvorov** [comment]
> @dimonchik-suvorov you should be able to reproduce the CI failures locally using breeze.

I'm trying :D

File: airflow/jobs/backfill_job.py (hunk 1 above)
Comment (ashb) [reply airflow/jobs/backfill_job.py]: YAGNI is my general rule -- keep the change as small as sensible for now, if we need it in the future then we do it then.

File: tests/jobs/test_scheduler_job.py [hunk 2]
```diff
@@ -2364,9 +2370,6 @@ def test_dagrun_root_fail_unfinished(self):
         )
         self.null_exec.mock_task_fail(dag_id, 'test_dagrun_fail', DEFAULT_DATE)
 
-        with pytest.raises(AirflowException):
```
Comment (dimonchik-suvorov) [inline tests/jobs/test_scheduler_job.py:1877]: Current impl doesn't rise any exception anyway

**Comment by dimonchik-suvorov** [comment]
@kaxil @XD-DENG @ashb @houqp guys, it's ready for review, I just don't know what to do with failing build... I didn't touch the flake part and I'm not sure what is `black` and why it failed.

**Comment by ashb** [comment]
@dimonchik-suvorov rebase to latest main and that should be fixed

**Comment by dimonchik-suvorov** [comment]
> @dimonchik-suvorov rebase to latest main and that should be fixed

I just did that... will try one more time

**Comment by dimonchik-suvorov** [comment]
Hi @kaxil, I've rebased from the upstream `main` branch again. PR is ready for review :)

File: tests/jobs/test_scheduler_job.py [hunk 3]
```diff
@@ -1846,6 +1854,10 @@ def test_dagrun_success(self):
             dagrun_state=State.SUCCESS,
         )
 
+    @unittest.skip(
+        "Hackish way to test scheduler which doesn't work anymore because the BackfillJob "
+        "doesn't interfere with 'running' DagRuns started by any other job type"
+    )
```
Comment (ashb) [inline tests/jobs/test_scheduler_job.py:1860]: You can't skip this test. Instead you will have to rewrite the `evaluate_dagrun` function to not use backfil.

**Review by ashb — CHANGES_REQUESTED** [review]

**Pull request closed without merging**